R/functions-range.R

```diff
@@ -219,3 +220,173 @@ pl__datetime_ranges <- function(
     )
   })
 }
+
+#' Generate a range of integers
+#'
+#' @inheritParams rlang::check_dots_empty0
+#' @param start Start of the range (inclusive). Defaults to 0.
+#' @param end End of the range (exclusive). If `NULL` (default), the value
+#' of `start` is used and `start` is set to 0.
+#' @param step Step size of the range.
+#' @param dtype Data type of the range.
+#'
+#' @inherit as_polars_expr return
+#' @examples
+#' pl$select(int = pl$int_range(0, 3))
+#'
+#' # end can be omitted for a shorter syntax.
+#' pl$select(int = pl$int_range(3))
+#'
+#' # Generate an index column by using int_range in conjunction with len().
+#' df <- pl$DataFrame(a = c(1, 3, 5), b = c(2, 4, 6))
+#' df$select(
+#'   index = pl$int_range(pl$len(), dtype = pl$UInt32),
+#'   pl$all()
+#' )
+pl__int_range <- function(
+    start = 0,
+    end = NULL,
+    step = 1,
+    ...,
+    dtype = pl$Int64) {
+  wrap({
+    check_dots_empty0(...)
+    check_polars_dtype(dtype)
+    if (is.null(end)) {
+      end <- start
+      start <- 0
+    }
+    int_range(
+      as_polars_expr(start)$`_rexpr`,
+      as_polars_expr(end)$`_rexpr`,
+      step,
+      dtype$`_dt`
+    )
+  })
+}
+
+#' Generate a range of integers for each row of the input columns
+#'
+#' @inheritParams pl__int_range
+#' @inherit as_polars_expr return
+#'
+#' @examples
+#' df <- pl$DataFrame(start = c(1, -1), end = c(3, 2))
+#' df$with_columns(int_range = pl$int_ranges("start", "end"))
+#'
+#' # end can be omitted for a shorter syntax$
+#' df$select("end", int_range = pl$int_ranges("end"))
+pl__int_ranges <- function(
+    start = 0,
+    end = NULL,
+    step = 1,
+    ...,
+    dtype = pl$Int64) {
+  wrap({
+    check_dots_empty0(...)
+    check_polars_dtype(dtype)
+    if (is.null(end)) {
+      end <- start
+      start <- 0
+    }
+    int_ranges(
+      as_polars_expr(start)$`_rexpr`,
+      as_polars_expr(end)$`_rexpr`,
+      as_polars_expr(step)$`_rexpr`,
+      dtype$`_dt`
+    )
+  })
+}
+
+#' Generate a time range
+#'
+#' @inheritParams rlang::check_dots_empty0
+#' @inheritParams pl__datetime_range
+#' @param start Lower bound of the time range. If omitted, defaults to
+#' `00:00:00.000`.
+#' @param end Upper bound of the time range. If omitted, defaults to
+#' `23:59:59.999`
+#' @param interval Interval of the range periods, specified as a [difftime]
+#' or using the Polars duration string language (see details).
+#'
+#' @inheritSection polars_duration_string Polars duration string language
+#' @inherit as_polars_expr return
+#' @examplesIf requireNamespace("hms", quietly = TRUE)
+#' pl$select(
+#'   time = pl$time_range(
+#'     start = hms::parse_hms("14:00:00"),
+#'     interval = as.difftime("3:15:00")
+#'   )
+#' )
+pl__time_range <- function(
+    start = NULL,
+    end = NULL,
```

Comment:
I feel there is no need to allow `NULL` here.
I think `start = pl$lit(0)$cast(pl$Time)` makes more sense.

Reply:
I disagree, I think:
* it would blur the fact that it's possible to pass something that is not an Expr as `start` and `end`
* for consistency we'd need to put `end = pl$lit(86399999999999)$cast(pl$Time)`, which is very unclear
* we'd need to check whether `start` is `NULL` anyway so it wouldn't save lines of code

(This also follows py-polars API btw)

Reply:
I understand. Could you use the `%||%` operator in that case to match the general pattern in R?

Reply:
Done